mcs/class/System/Microsoft.CSharp/CSharpCodeGenerator.cs

```diff
@@ -93,18 +93,16 @@ private CompilerResults FromFileBatch (CompilerParameters options, string[] file
 
 			mcs.StartInfo.CreateNoWindow=true;
 			mcs.StartInfo.UseShellExecute=false;
-			mcs.StartInfo.RedirectStandardOutput=true;
+			mcs.StartInfo.WindowStyle = ProcessWindowStyle.Hidden;
 			mcs.StartInfo.RedirectStandardError=true;
 			mcs.ErrorDataReceived += new DataReceivedEventHandler ((sender, args) => {
-				if (args.Data != null) {
-					mcsOutMutex.WaitOne ();
+				if (args.Data != null)
 					results.Output.Add (args.Data);
```

Comment:
I looked into some of our homegrown test-runner implementations which also do Process.Start() and they `lock` inside of the event handler before appending to the stringbuilder, maybe add that here too?

Reply:
Talking with @kg on Slack this might be the actual underlying cause of the bug - if an exception happens on the event handler the async read breaks.

Reply:
The lock inside the handler is not needed - it won't be called simultaneously on several threads.